pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +238,267 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+						Timestamp:   state.CommitRecTimestamp,
+						LeaderEpoch: state.Commit.LeaderEpoch,
+					}
+				} else if commitRec == nil {
+					// The commit offset should be the last record that was fully processed and not the first record that was not fully processed.
+					commitRec = lastRec
+				}
+			}
+			lastRec = rec
+		}
+	}
+
+	// Nothing was consumed from Kafka at all.
+	if firstRec == nil {
+		level.Info(logger).Log("msg", "no records were consumed")
+		return state, nil
+	}
+
+	// No records were processed for this cycle.
+	if lastRec == nil {
+		level.Info(logger).Log("msg", "nothing to commit due to first record fetched is from next cycle", "first_rec_offset", firstRec.Offset)
+		return state, nil
+	}
+
+	// All samples in all records were processed. We can commit the last record's offset.
+	if commitRec == nil {
+		commitRec = lastRec
+	}
+
+	var err error
+	numBlocks, err = builder.CompactAndUpload(ctx, b.uploadBlocks)
+	if err != nil {
+		return state, err
+	}
+
+	// We should take the max of last seen offsets. If the partition was lagging due to some record not being processed
+	// because of a future sample, we might be coming back to the same consume cycle again.
+	lastSeenOffset := max(lastRec.Offset, state.LastSeenOffset)
+	// Take the max of block max times because of the same reasons above.
+	lastBlockEnd := blockEnd
+	if lastBlockEnd.Before(state.LastBlockEnd) {
+		lastBlockEnd = state.LastBlockEnd
+	}
+	commit := kadm.Offset{
+		Topic:       commitRec.Topic,
+		Partition:   commitRec.Partition,
+		At:          commitRec.Offset + 1, // offset+1 means everything up to (including) the offset was processed
+		LeaderEpoch: commitRec.LeaderEpoch,
+		Metadata:    marshallCommitMeta(commitRec.Timestamp.UnixMilli(), lastSeenOffset, lastBlockEnd.UnixMilli()),
+	}
+	newState := partitionState{
+		Lag:                state.Lag - (commit.At - firstRec.Offset), // the new lag is the distance between fully processed offsets
+		Commit:             commit,
+		CommitRecTimestamp: commitRec.Timestamp,
+		LastSeenOffset:     lastSeenOffset,
+		LastBlockEnd:       lastBlockEnd,
+	}
+	if err := b.commitState(ctx, logger, b.cfg.Kafka.ConsumerGroup, newState); err != nil {
+		return state, err
+	}
+
+	return newState, nil
+}
+
+func (b *BlockBuilder) commitState(ctx context.Context, logger log.Logger, group string, state partitionState) error {
+	offsets := make(kadm.Offsets)
+	offsets.Add(state.Commit)
+	if err := kadm.NewClient(b.kafkaClient).CommitAllOffsets(ctx, group, offsets); err != nil {
+		return fmt.Errorf("commit with partition %d, offset %d: %w", state.Commit.Partition, state.Commit.At, err)
+	}
+
+	level.Debug(logger).Log("msg", "successfully committed to Kafka", "offset", state.Commit.At)
+
+	return nil
+}
+
+func (b *BlockBuilder) uploadBlocks(ctx context.Context, tenantID, dbDir string, blockIDs []string) error {
```

Comment:
I would wrap it into another function `uploadBlocksWithRetries()`. I don't want to re-run the whole cycle just because of a transient networking issue.